core/node/node_framework/src/implementations/layers/consensus.rs

```diff
@@ -0,0 +1,176 @@
+                    action_queue_sender,
+                };
+                context.add_task(Box::new(task));
+            }
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+pub struct MainNodeConsensusTask {
+    config: MainNodeConfig,
+    pool: ConnectionPool<Core>,
+}
+
+#[async_trait::async_trait]
+impl Task for MainNodeConsensusTask {
+    fn name(&self) -> &'static str {
+        "consensus"
+    }
+
+    async fn run(self: Box<Self>, mut stop_receiver: StopReceiver) -> anyhow::Result<()> {
+        // We instantiate the root context here, since the consensus task is the only user of the
+        // structured concurrency framework (`MainNodeConsensusTask` and `FetcherTask` are considered mutually
+        // exclusive).
+        // Note, however, that awaiting for the `stop_receiver` is related to the root context behavior,
+        // not the consensus task itself. There may have been any number of tasks running in the root context,
+        // but we only need to wait for stop signal once, and it will be propagated to all child contexts.
+        let root_ctx = ctx::root();
+        scope::run!(&root_ctx, |ctx, s| async move {
+            s.spawn_bg(consensus::era::run_main_node(ctx, self.config, self.pool));
+            let _ = stop_receiver.0.wait_for(|stop| *stop).await?;
+            Ok(())
+        })
+        .await
+    }
+}
+
+#[derive(Debug)]
+pub struct FetcherTask {
+    config: Option<(consensus::Config, consensus::Secrets)>,
+    pool: ConnectionPool<Core>,
+    main_node_client: Arc<dyn MainNodeClient>,
+    sync_state: SyncState,
+    action_queue_sender: ActionQueueSender,
+}
+
+#[async_trait::async_trait]
+impl Task for FetcherTask {
+    fn name(&self) -> &'static str {
+        "consensus_fetcher"
+    }
+
+    async fn run(self: Box<Self>, mut stop_receiver: StopReceiver) -> anyhow::Result<()> {
+        // We instantiate the root context here, since the consensus task is the only user of the
+        // structured concurrency framework (`MainNodeConsensusTask` and `FetcherTask` are considered mutually
+        // exclusive).
+        // Note, however, that awaiting for the `stop_receiver` is related to the root context behavior,
+        // not the consensus task itself. There may have been any number of tasks running in the root context,
+        // but we only need to wait for stop signal once, and it will be propagated to all child contexts.
+        let root_ctx = ctx::root();
+        scope::run!(&root_ctx, |ctx, s| async {
+            s.spawn_bg(async move {
+                zksync_core::consensus::era::run_fetcher(
+                    &root_ctx,
+                    self.config,
+                    self.pool,
+                    self.sync_state,
+                    self.main_node_client,
+                    self.action_queue_sender,
+                )
+                .await
+            });
+            ctx.wait(stop_receiver.0.wait_for(|stop| *stop)).await??;
```

Comment:
fyi, if stop_receiver gets disconnected, then `FetcherTask::run()` will return an error. Is that intended?

Reply:
I think it's OK. `stop_sender` per framework rules may only be dropped after either all tasks have exited (meaning that `FetcherTask` is not affected) or the task has ignored the stop signal (not the case here).
So to me, it looks like it can only return an error if there is a bug in the framework, which sounds reasonable.